### PR TITLE
feat(infra): livedata flatten

### PR DESCRIPTION
**Pull request by EYHN**

### CODE REVIEW

**Comment by EYHN** [comment]
* **#6083** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6083?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#6084](https://github.com/toeverything/AFFiNE/pull/6084) <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6084?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#6091](https://github.com/toeverything/AFFiNE/pull/6091) <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6091?utm_source=stack-comment-icon" target="_blank">Graphite</a>) 👈
* `canary`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @EYHN and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/common/infra/src/livedata/index.ts [hunk 1]
```diff
@@ -268,6 +269,38 @@ export class LiveData<T = unknown> implements InteropObservable<T> {
...
+   *
+   * ```
+   * new LiveData(new LiveData(0)).flat() // LiveData<number>
+   * ```
+   *
+   * ```
+   * new LiveData([new LiveData(0)]).flat() // LiveData<number[]>
+   * ```
+   */
+  flat(): T extends LiveData<infer P>
```
Comment (forehalo) [inline packages/common/infra/src/livedata/index.ts]: should it be auto flatten like promise?

File: packages/common/infra/src/livedata/index.ts [hunk 2]
```diff
@@ -268,6 +269,38 @@ export class LiveData<T = unknown> implements InteropObservable<T> {
...
+    ? LiveData<P>
+    : T extends LiveData<infer P>[]
+      ? LiveData<P[]>
+      : never {
+    return LiveData.from(
+      this.pipe(
+        switchMap(v => {
+          if (v instanceof LiveData) {
+            return v as LiveData<any>;
+          } else if (v instanceof Array) {
```
Comment (pengx17) [inline packages/common/infra/src/livedata/index.ts]: I think `Array.isArray` is preferred over `instanceof Array`

File: packages/common/infra/src/livedata/index.ts (hunk 1 above)
Comment (EYHN) [reply packages/common/infra/src/livedata/index.ts]: I tried to implement auto flatten, and it works, but some cases are not very clear. You can still define `LiveData<LiveData<T>>` just like `Promise<Promise<T>>` which is legal in typescript but does not actually exist.

File: packages/common/infra/src/livedata/index.ts (hunk 2 above)
Comment (EYHN) [reply packages/common/infra/src/livedata/index.ts]: done

**Review by pengx17 — APPROVED** [review]

**Pull request merged**